keps/sig-node/20190925-enable-seccomp-by-default.md

```diff
@@ -0,0 +1,327 @@
+
+```
+net.core.message_burst = 10
+net.core.message_cost = 5
+```
+
+Users can adjust the verbosity of their logs on demand, although the default settings are quite safe.
+
+
+#### Cost vs Benefit
+
+The performance overhead of auditing syscalls that not whitelisted are negligible, once the cost of enabling seccomp is taken into account. 
+
+Kubernetes should take the same approach taken by docker of enabling both AppArmor and Seccomp by default. This would make any overhead added by such features part of the cost for running secure container workloads. 
+
+Providing users with sensible default security settings outweights the performance impact incurred, specially when users are given the ability to opt-out if their prefer to do so.
+
+
+### User Stories
+
+#### Story 1
+As a user and administrator, I want to be able to assess what pods within my cluster would be affected by applying more restrictive seccomp profiles.
+
+#### Story 2
+As a user, I want to be able to assess whether my pods will be affected by more restrictive seccomp profiles.
+
+#### Story 3
+As a user and administrator, I want Kubernetes to audit all potentially dangerous system calls from my containers by default.
+
+
+
+## Implementation Details
+
+### 1. Audit mode support (Details)
+
+Kubernetes will continue to be unaware of downstream support. If a user tries to apply a seccomp profile containing unsupported actions (i.e. `SCMP_ACT_LOG`) today, the lower level dependencies (i.e. runC) will return an error. As a result, the container won't be able to start. We are proposing no changes to this behaviour.
+
+
+### 2. Built-in profiles (Details)
+
+The internal built-in profiles will be implemented in golang, not allowing users to amend them. It will be based off the implementation in [docker](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) and [containerd](https://github.com/containerd/containerd/blob/master/contrib/seccomp/seccomp_default.go).
+
+The new built-in profiles would be mapped to an additional Kubernetes profile type:
+```
+const (
+    SeccompProfileUnconfined SeccompProfileType = "Unconfined"
+    SeccompProfileRuntime    SeccompProfileType = "Runtime"
+    SeccompProfileLocalhost  SeccompProfileType = "Localhost"
+    SeccompProfileKubernetes SeccompProfileType = "Kubernetes"
+)
+```
+
+Profiles will be passed as a serialised json object inside dockerOpt objects, in the same way that it is done currently for file based profiles:
+```
+jsonSeccomp, _ := json.Marshal(profile.Spec)
+return []dockerOpt{{"seccomp", string(jsonSeccomp), seccompProfileName}}, nil
+```
+_Needs confirmation as to whether this would also work for non-docker runtime implementations._
+
+
+Built-in profiles will act in the same way as custom profiles, therefore, trying to apply any of the audit built-in profiles in nodes which do not support `SCMP_ACT_LOG` (due to kernel or dependencies version) will cause containers to fail to start.
```

Comment:
The table above (#2-built-in-profiles-details) mentions a fallback profile. How does that work in this case?